### PR TITLE
Feature/sync game score

### PR DESCRIPTION
**Pull request by kangpl**

### CODE REVIEW

File: app/src/androidTest/java/ch/epfl/sdp/database/firebase/ClientMockDatabaseAPI.java [hunk 1]
```diff
@@ -24,17 +22,7 @@ public void listenToGameStart(OnValueReadyCallback<CustomResult<Void>> onValueRe
...
-    }
-
-    @Override
-    public void addItemBoxesListener(OnValueReadyCallback<CustomResult<List<ItemBoxForFirebase>>> onValueReadyCallback) {
-
-    }
-
-    @Override
-    public void addUserHealthPointsListener(OnValueReadyCallback<CustomResult<Double>> onValueReadyCallback) {
+    public void addCollectionListerner(Object entityType, OnValueReadyCallback<CustomResult<List<Object>>> onValueReadyCallback) {
```
Comment (varathar53) [inline app/src/androidTest/java/ch/epfl/sdp/database/firebase/ClientMockDatabaseAPI.java:25]: Good to put merge like addEnemyList..addItemBoxesListen into one. :)

File: app/src/main/java/ch/epfl/sdp/leaderboard/GeneralLeaderboardActivity.java [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+
+        addGeneralGameScoreListener();
+    }
+
+    public void setupChampions(List<LeaderboardEntity> players) {
+        TextView tv_username = null;
+        TextView tv_healthpoint = null;
+
+        if (players != null) {
+            for(int i = 0; i < players.size() && i < 3; i++) {
```
Comment (varathar53) [inline app/src/main/java/ch/epfl/sdp/leaderboard/GeneralLeaderboardActivity.java]: its too long this if statement make it smaller.

File: app/src/main/java/ch/epfl/sdp/database/firebase/api/ClientFirestoreDatabaseAPI.java [hunk 3]
```diff
@@ -39,49 +39,33 @@ public void listenToGameStart(OnValueReadyCallback<CustomResult<Void>> onValueRe
...
-                    else {
-                        List<ItemBoxForFirebase> itemBoxForFirebaseList = new ArrayList<>();
-                        for (DocumentChange documentChange : querySnapshot.getDocumentChanges()) {
-                            itemBoxForFirebaseList.add(documentChange.getDocument().toObject(ItemBoxForFirebase.class));
-                        }
-                        onValueReadyCallback.finish(new CustomResult<>(itemBoxForFirebaseList, true, null));
-                    }
-                });
-    }
+    public void addCollectionListerner(Object entityType, OnValueReadyCallback<CustomResult<List<Object>>> onValueReadyCallback) {
```
Comment (Zeavee) [inline app/src/main/java/ch/epfl/sdp/database/firebase/api/ClientFirestoreDatabaseAPI.java:43]: Nice!

File: app/src/main/java/ch/epfl/sdp/database/firebase/api/CommonFirestoreDatabaseAPI.java [hunk 4]
```diff
@@ -74,11 +62,39 @@ public void selectLobby(OnValueReadyCallback<CustomResult<Void>> onValueReadyCal
...
+                .addOnSuccessListener(queryDocumentSnapshots -> {
+                    List<PlayerForFirebase> playerForFirebases = new ArrayList<>();
+                    for (QueryDocumentSnapshot document : queryDocumentSnapshots) {
+                        playerForFirebases.add(document.toObject(PlayerForFirebase.class));
+                    }
+                    onValueReadyCallback.finish(new CustomResult<>(playerForFirebases, true, null));
+                }).addOnFailureListener(e -> onValueReadyCallback.finish(new CustomResult<>(null, false, e)));
+    }
+
+    @Override
```
[the remark this answers is not in the document]
Comment (Zeavee) [reply app/src/main/java/ch/epfl/sdp/database/firebase/api/CommonFirestoreDatabaseAPI.java:85]: I don't know if it is possible to fix this

File: app/src/main/java/ch/epfl/sdp/map/GoogleLocationFinder.java [hunk 5]
```diff
@@ -29,10 +30,11 @@ public GoogleLocationFinder(LocationManager locationManager) {
         locationListener = new LocationListener() {
             @Override
             public void onLocationChanged(Location latestLocation) {
-                currentLocation = latestLocation;
-                PlayerManager.getInstance().getCurrentUser().setLocation(new GeoPoint(latestLocation.getLongitude(), latestLocation.getLatitude()));
-                PlayerManager.getInstance().getCurrentUser().setLocation(PlayerManager.getInstance().getCurrentUser().getLocation());
-                requestUpdatePosition();
+                if (Math.floor(latestLocation.getLatitude() * 100) / 100 != 37.42 || Math.ceil(latestLocation.getLongitude() * 100) / 100 != -122.08) {
```
Comment (Zeavee) [inline app/src/main/java/ch/epfl/sdp/map/GoogleLocationFinder.java:33]: What is this formula?

File: app/src/main/java/ch/epfl/sdp/map/MapsActivity.java [hunk 6]
```diff
@@ -161,6 +166,11 @@ public void onMapReady(GoogleMap googleMap) {
                     Toast.makeText(MapsActivity.this, fetchUserRes.getException().getMessage(), Toast.LENGTH_LONG).show();
                 }
             });
+        } else {
+            // reDisplay the itemBox when resume the map
+            for(ItemBox itemBox: ItemBoxManager.getInstance().getItemBoxes().values()) {
```
Comment (Zeavee) [inline app/src/main/java/ch/epfl/sdp/map/MapsActivity.java:171]: Oh, I see

File: app/src/main/java/ch/epfl/sdp/map/GoogleLocationFinder.java (hunk 5 above)
Comment (kangpl) [reply app/src/main/java/ch/epfl/sdp/map/GoogleLocationFinder.java:33]: Sometimes, the player's location will change to this wired GeoPoint(which is the base of Google in the USA) automatically even the player stays in the fixed position. It seems like a problem from the emulator, so I filtered this location.

File: app/src/main/java/ch/epfl/sdp/leaderboard/GeneralLeaderboardActivity.java (hunk 2 above)
Comment (kangpl) [reply app/src/main/java/ch/epfl/sdp/leaderboard/GeneralLeaderboardActivity.java]: fixed :)

File: app/src/main/java/ch/epfl/sdp/map/GoogleLocationFinder.java (hunk 5 above)
Comment (Zeavee) [reply app/src/main/java/ch/epfl/sdp/map/GoogleLocationFinder.java:33]: Oh, I see

**Comment by kangpl** [comment]
> Hey Peilin is it not possible to increase a bit the test coverage (so that master at least remains at the same coverage). Because Rafael (or Sen's merge?) already got the coverage to 81% and this one will probably decrease it to below 80%

Yes, and now the total coverage is 81.9%

**Pull request merged**